app/partials/app.cjsx

```diff
@@ -3,15 +3,81 @@ auth = require 'panoptes-client/lib/auth'
 IOStatus = require './io-status'
 MainHeader = require './main-header'
 MainFooter = require './main-footer'
+GeordiClient = require 'zooniverse-geordi-client'
 {generateSessionID} = require '../lib/session'
 
+class GeordiLogger
+  constructor: (@state, @geordi) ->
+
+  keys: {}
+
+  geordi: null
+
+  instance: =>
+    @geordi = @geordi || @makeGeordi @keys?.projectToken
+
+  makeGeordi: (projectSlug) ->
+    new GeordiClient
+      server: @state?.env
+      projectToken: projectSlug || ''
+      zooUserIDGetter: () => @state.user?.id
+      subjectGetter: () => @keys?.subjectID
+
+  makeHandler: (defType) ->
+    instance = @instance
+    (eventData, eventType) ->
+        eventType = defType if typeof linkType isnt 'string'
+        instance()?.logEvent
+          type: eventType
+          data: "\"#{eventData}\""
+
+  logEvent: (logEntry) ->
+    geordi = @instance()
+    newEntry = Object.assign {}, logEntry, @keys
+    geordi?.logEvent newEntry
+    console.log 'No logger available for event ', JSON.stringify(logEntry) unless geordi?.logEvent
+
+  remember: (keyVals) ->
+    rebuild = keyVals?.projectToken? && (keyVals?.projectToken != @geordi?.projectToken)
+    @geordi = @makeGeordi keyVals.projectToken if rebuild
+    @keys = Object.assign {}, @keys, keyVals
+
+  forget: (forgetKeys) ->
+    rebuild = false
+    for key in forgetKeys
+      rebuild = true if key == 'projectToken'
+      delete @keys[key]
+
+    @geordi = @makeGeordi '' if rebuild
+
 module.exports = React.createClass
   displayName: 'PanoptesApp'
 
+  geordiLogger: null
```

Comment:
Is Geordi like a Star Trek thing?

On Thursday, June 16, 2016, Alex Bowyer <notifications@github.com> wrote:

> In app/partials/app.cjsx
> <https://github.com/zooniverse/Panoptes-Front-End/pull/2590#discussion_r67374209>
> :
>
> >  module.exports = React.createClass
> >    displayName: 'PanoptesApp'
> >
> > +  geordiLogger: null
>
> @will <https://github.com/will> Something simple like "An instance of the
> geordi client that can be used by the consumer of this class to log events
> to the Geordi server and google analytics"
>
> Actually could probably go even simpler than that, it's a bit wordy. I
> just think since the class is effectively offering an interface to the rest
> of the codebase it warrants at least a short comment.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/zooniverse/Panoptes-Front-End/pull/2590/files/f66d8c6b812f33748f34b69abb3c7c9fd959b215#r67374209>,
> or mute the thread
> <https://github.com/notifications/unsubscribe/AAAHtUJ7o3f3yURD_JUzmVkY2kEB7N8Qks5qMXT3gaJpZM4IfsJ9>
> .
>